File: backend/app/services/scoring_engine.py
"""Deterministic Career Readiness Scoring Engine for ORBIT.

Pure mathematical implementation with zero LLM/hallucinated numerical values.
Strictly bounded to [0, 100] with fully explainable 4-pillar formulation.

Formula:
  base_score = 0.35 * skill_alignment + 0.25 * code_evidence + 0.20 * production_hygiene + 0.20 * architecture_depth
  overall_score = min(100, max(0, round(base_score)))
"""
import math
from typing import Dict, Any, List, Optional, Tuple, Literal
from pydantic import BaseModel, Field


# Approved 4-state Evidence Model
EvidenceState = Literal["VERIFIED", "PARTIAL", "SELF_REPORTED", "UNAVAILABLE"]


class PillarBreakdown(BaseModel):
    skillAlignment: int = Field(ge=0, le=100)
    codeEvidence: int = Field(ge=0, le=100)
    productionHygiene: int = Field(ge=0, le=100)
    architectureDepth: int = Field(ge=0, le=100)


class SkillGapEvaluation(BaseModel):
    skill: str
    category: str
    importance: str  # "Required" | "Preferred"
    candidateEvidence: str
    repoCount: int = 0
    hasCert: bool = False
    evidenceState: EvidenceState
    gapLevel: str       # "Low" | "Medium" | "High"
    evidenceStrength: int = Field(ge=0, le=100)
    impactScore: str
    actionItem: str


class ReadinessEvaluationResult(BaseModel):
    overall_score: int = Field(ge=0, le=100)
    base_score: int = Field(ge=0, le=100)
    pillars: PillarBreakdown
    gap_matrix: List[SkillGapEvaluation]
    completed_milestones: List[str]
    status_badge: str
    rating: str


# ====================================================================
# Pillar 1: Skill Alignment Calculation (Weight: 35%)
# ====================================================================
def calculate_skill_alignment(
    job_skills: List[Dict[str, Any]],
    candidate_skills: List[Dict[str, Any]],
    candidate_repos: List[Dict[str, Any]],
    completed_milestones: Optional[List[str]] = None
) -> Tuple[float, List[SkillGapEvaluation]]:
    """Calculates deterministic Skill Alignment score [0, 100] using Phase 4 JD weights.
    
    Strictly preserves the 4 approved evidence states:
      1. VERIFIED (>= 1 verified repos with matching tech in code):
         - >= 2 verified repos: E_i = 1.00 (evidenceStrength = min(100, 75 + repoCount * 5))
         - 1 verified repo: E_i = 0.85 (evidenceStrength = 75)
      2. PARTIAL (Valid certification matching skill, 0 repos):
         - E_i = 0.55 (evidenceStrength = 50)
      3. SELF_REPORTED (Resume mention only, 0 verified repos):
         - E_i = 0.30 (evidenceStrength = 25)
      4. UNAVAILABLE (Absent in analyzed repositories and resume):
         - E_i = 0.00 (evidenceStrength = 0)
    """
    if not job_skills:
        return 100.0, []

    total_weight = sum(s.get("weight", 0.0) for s in job_skills)
    if total_weight <= 0:
        total_weight = 1.0

    skill_alignment_score = 0.0
    gap_matrix = []

    for req_skill in job_skills:
        skill_name = req_skill["name"]
        skill_weight = req_skill.get("weight", 1.0 / len(job_skills))
        importance = req_skill.get("importance", "Required")
        category = req_skill.get("category", "General")

        # 1. Check candidate repository code proof (Real Evidence)
        matching_repos = [
            r for r in candidate_repos
            if any(t.lower() == skill_name.lower() for t in r.get("tech_stack", [])) or
               (r.get("language", "").lower() == skill_name.lower())
        ]
        repo_count = len(matching_repos)

        # 2. Check resume claims and certifications
        cand_skill_entry = next(
            (s for s in candidate_skills if s.get("name", "").lower() == skill_name.lower()),
            None
        )
        has_cert = cand_skill_entry.get("certified", False) if cand_skill_entry else False
        cert_name = cand_skill_entry.get("certName", "Certification") if cand_skill_entry else ""
        resume_mentions = cand_skill_entry.get("resumeMentions", 0) if cand_skill_entry else 0

        # 3. Deterministic 4-State Evidence Assignment
        if repo_count >= 2:
            evidence_state: EvidenceState = "VERIFIED"
            evidence_factor = 1.00
            evidence_str = f"{repo_count} GitHub repositories with verified code"
            gap_level = "Low"
            evidence_strength = min(100, 75 + repo_count * 5)
            impact_str = "Verified Strength"
            action_item = "Maintain test coverage and active documentation"
        elif repo_count == 1:
            evidence_state: EvidenceState = "VERIFIED"
            evidence_factor = 0.85
            evidence_str = "1 GitHub repository (Verified in Code)"
            gap_level = "Low"
            evidence_strength = 75
            impact_str = "Verified Match"
            action_item = "Expand test coverage and module depth in repository"
        elif has_cert:
            evidence_state = "PARTIAL"
            evidence_factor = 0.55
            evidence_str = f"Certified ({cert_name}) — 0 code repositories"
            gap_level = "Medium"
            evidence_strength = 50
            impact_str = "+8 pts"
            action_item = f"Build 1 production repository demonstrating {skill_name}"
        elif resume_mentions > 0:
            evidence_state = "SELF_REPORTED"
            evidence_factor = 0.30
            evidence_str = f"Self-reported on resume ({resume_mentions}x) — 0 code repositories"
            gap_level = "High" if importance == "Required" else "Medium"
            evidence_strength = 25
            impact_str = "+12 pts" if importance == "Required" else "+6 pts"
            action_item = f"Create GitHub repository providing code proof for {skill_name}"
        else:
            evidence_state = "UNAVAILABLE"
            evidence_factor = 0.00
            evidence_str = "Not found in analyzed technical evidence"
            gap_level = "High" if importance == "Required" else "Medium"
            evidence_strength = 0
            impact_str = "+15 pts" if importance == "Required" else "+5 pts"
            action_item = f"Complete a project blueprint to build and deploy {skill_name}"

        # Weighted Contribution
        skill_alignment_score += (skill_weight / total_weight) * (evidence_factor * 100.0)

        gap_matrix.append(SkillGapEvaluation(
            skill=skill_name,
            category=category,
            importance=importance,
            candidateEvidence=evidence_str,
            repoCount=repo_count,
            hasCert=has_cert,
            evidenceState=evidence_state,
            gapLevel=gap_level,
            evidenceStrength=evidence_strength,
            impactScore=impact_str,
            actionItem=action_item
        ))

    return min(100.0, max(0.0, skill_alignment_score)), gap_matrix


# ====================================================================
# Pillar 2: GitHub Code Evidence (Weight: 25%)
# ====================================================================
def calculate_code_evidence(candidate_repos: List[Dict[str, Any]]) -> float:
    """Calculates GitHub Code Evidence score [0, 100] from real verifiable code telemetry.
    
    Formulation (Max 100 pts):
      1. Repo Count Volume: min(40, N_repos * 10) -> Max 40 pts
      2. Complexity Score: min(35, (avg_complexity / 10.0) * 35) -> Max 35 pts
      3. Commit Cadence (365d): min(15, (commits_365d / 150) * 15) -> Max 15 pts (Supporting signal)
      4. Documentation Depth: min(10, (avg_readme_words / 800) * 10) -> Max 10 pts
      * Stars = 0 pts (Social proof only, zero impact on code evidence).
    """
    if not candidate_repos:
        return 0.0

    n_repos = len(candidate_repos)
    repo_volume_score = min(40.0, n_repos * 10.0)

    avg_complexity = sum(r.get("complexity_score", 5.0) for r in candidate_repos) / n_repos
    complexity_score = min(35.0, (avg_complexity / 10.0) * 35.0)

    total_commits_year = sum(r.get("commits_last_year", 0) for r in candidate_repos)
    commit_score = min(15.0, (total_commits_year / 150.0) * 15.0)

    # Documentation depth: uses readme_words (default 800 words = 10 pts)
    avg_readme = sum(r.get("readme_words", 800.0) for r in candidate_repos) / n_repos
    doc_score = min(10.0, (avg_readme / 800.0) * 10.0)

    total = repo_volume_score + complexity_score + commit_score + doc_score
    return min(100.0, max(0.0, total))


# ====================================================================
# Pillar 3: Production & DevOps Hygiene (Weight: 20%)
# ====================================================================
def calculate_production_hygiene(candidate_repos: List[Dict[str, Any]]) -> float:
    """Calculates Production Hygiene score [0, 100].
    
    Formulation (Max 100 pts):
      1. Testing Component (Max 35 pts):
         - Verified Positive Coverage (test_coverage > 0.0): (test_coverage / 100.0) * 35.0
         - Verified Zero Coverage (test_coverage == 0.0): 5.0 pts (test files present, 0% passing coverage)
         - Unavailable Coverage (test_coverage is None):
           - If has_tests == True: 17.5 pts (baseline test presence credit without assuming coverage)
           - If has_tests == False: 0.0 pts
      2. Docker / Containerization Presence (Max 35 pts):
         - (N_docker / N_repos) * 35.0
      3. Automated CI/CD Workflows (Max 30 pts):
         - (N_ci / N_repos) * 30.0
    """
    if not candidate_repos:
        return 0.0

    n_repos = len(candidate_repos)
    
    testing_points = 0.0
    for r in candidate_repos:
        cov = r.get("test_coverage")
        has_tests = r.get("has_tests", False)

        if cov is not None:
            if cov > 0.0:
                testing_points += (min(100.0, float(cov)) / 100.0) * 35.0
            else:
                testing_points += 5.0  # Verified 0%
        else:
            if has_tests:
                testing_points += 17.5  # Presence credit only
            else:
                testing_points += 0.0

    test_score = testing_points / n_repos

    n_docker = sum(1 for r in candidate_repos if r.get("has_docker", False))
    n_ci = sum(1 for r in candidate_repos if r.get("has_ci", False))

    docker_score = (n_docker / n_repos) * 35.0
    ci_score = (n_ci / n_repos) * 30.0

    total = test_score + docker_score + ci_score
    return min(100.0, max(0.0, total))


# ====================================================================
# Pillar 4: Architecture & Engineering Structure Depth (Weight: 20%)
# ====================================================================
def calculate_architecture_depth(candidate_repos: List[Dict[str, Any]]) -> float:
    """Calculates Architecture & Engineering Structure Depth [0, 100].
    
    Strict Rules:
      - Stars do NOT increase architecture depth.
      - Certifications do NOT increase architecture depth.
      - Scored strictly on verifiable structural signals across repositories:
        1. Modular Layer Separation (has_modular_structure): +25 pts
        2. Database Integration & Migrations (has_db_integration): +20 pts
        3. Configuration Management (has_env_config): +15 pts
        4. Multi-Service / Containerized Infra (has_docker): +20 pts
        5. API Contracts & Specs (REST / FastAPI / Type schemas): +20 pts
    """
    if not candidate_repos:
        return 0.0

    has_modular = any(r.get("has_modular_structure", False) for r in candidate_repos)
    has_db = any(r.get("has_db_integration", False) for r in candidate_repos)
    has_config = any(r.get("has_env_config", False) for r in candidate_repos)
    has_docker = any(r.get("has_docker", False) for r in candidate_repos)
    has_api_contracts = any(
        "FastAPI" in r.get("tech_stack", []) or "REST APIs" in r.get("tech_stack", []) or "GraphQL" in r.get("tech_stack", [])
        for r in candidate_repos
    )

    score = 0.0
    if has_modular:
        score += 25.0
    if has_db:
        score += 20.0
    if has_config:
        score += 15.0
    if has_docker:
        score += 20.0
    if has_api_contracts:
        score += 20.0

    return min(100.0, max(0.0, score))


# ====================================================================
# Main Deterministic Scoring Orchestrator
# ====================================================================
def evaluate_readiness(
    candidate: Dict[str, Any],
    job: Dict[str, Any],
    candidate_repos: List[Dict[str, Any]],
    completed_milestones: Optional[List[str]] = None
) -> ReadinessEvaluationResult:
    """Evaluates candidate technical evidence against target job requirements.
    
    Final Score Bounds:
      base_score = 0.35 * skill_alignment + 0.25 * code_evidence + 0.20 * production_hygiene + 0.20 * architecture_depth
      overall_score = min(100, max(0, round(base_score)))
      
    Deterministic Rating Scale:
      - [85, 100]: High Readiness (Verified technical match across core role requirements)
      - [70, 84]:  Moderate Readiness (Satisfies primary requirements with minor gap closure needed)
      - [50, 69]:  Developing Readiness (Foundational skills present; needs production repository proof)
      - [0, 49]:   Emerging Readiness (Early stage; significant skill & architecture gaps)
    """
    milestones = completed_milestones or []
    all_job_skills = (job.get("required_skills", []) or []) + (job.get("preferred_skills", []) or [])
    candidate_skills = candidate.get("extracted_skills", []) or []

    # 1. Compute 4 Pillars
    skill_alignment_val, gap_matrix = calculate_skill_alignment(
        all_job_skills, candidate_skills, candidate_repos, milestones
    )
    code_evidence_val = calculate_code_evidence(candidate_repos)
    production_hygiene_val = calculate_production_hygiene(candidate_repos)
    architecture_depth_val = calculate_architecture_depth(candidate_repos)

    # 2. Base Score & Overall Score (100% Deterministic, strictly 0-100)
    base_score_float = (
        0.35 * skill_alignment_val +
        0.25 * code_evidence_val +
        0.20 * production_hygiene_val +
        0.20 * architecture_depth_val
    )
    overall_score = min(100, max(0, round(base_score_float)))

    # 3. Deterministic Rating Scale
    if overall_score >= 85:
        status_badge = "High Readiness"
        rating = "High Readiness"
    elif overall_score >= 70:
        status_badge = "Moderate Readiness"
        rating = "Moderate Readiness"
    elif overall_score >= 50:
        status_badge = "Developing Readiness"
        rating = "Developing Readiness"
    else:
        status_badge = "Emerging Readiness"
        rating = "Emerging Readiness"

    pillars = PillarBreakdown(
        skillAlignment=round(skill_alignment_val),
        codeEvidence=round(code_evidence_val),
        productionHygiene=round(production_hygiene_val),
        architectureDepth=round(architecture_depth_val)
    )

    return ReadinessEvaluationResult(
        overall_score=overall_score,
        base_score=overall_score,
        pillars=pillars,
        gap_matrix=gap_matrix,
        completed_milestones=milestones,
        status_badge=status_badge,
        rating=rating
    )

File: src/utils/scoringEngine.js
/**
 * ORBIT Deterministic Career Readiness Scoring Engine (Client-Side)
 * Strictly synchronized with backend Phase 5 mathematical formulation.
 */

export function calculateReadinessScore(candidate, job, completedMilestoneIds = []) {
  if (!candidate || !job) {
    return {
      overallScore: 0,
      baseScore: 0,
      pillars: {
        skillAlignment: 0,
        codeEvidence: 0,
        productionHygiene: 0,
        architectureDepth: 0
      },
      rating: "Emerging Readiness",
      statusBadge: "Emerging Readiness",
      strengths: [],
      criticalGaps: []
    };
  }

  // 1. Skill Alignment (Weight: 35%)
  const allJobSkills = [...(job.requiredSkills || []), ...(job.preferredSkills || [])];
  let totalSkillWeight = 0;
  let earnedSkillScore = 0;
  const criticalGaps = [];
  const strengths = [];

  allJobSkills.forEach(reqSkill => {
    const isRequired = reqSkill.importance === "Required";
    const skillWeight = reqSkill.weight || (isRequired ? 0.1 : 0.05);
    totalSkillWeight += skillWeight;

    // Check candidate repos
    const matchingRepos = (candidate.githubRepos || []).filter(r => 
      (r.techStack || r.tech_stack || []).some(t => t.toLowerCase() === reqSkill.name.toLowerCase()) ||
      (r.language || '').toLowerCase() === reqSkill.name.toLowerCase()
    );
    const repoCount = matchingRepos.length;

    // Check resume & certs
    const candidateSkill = (candidate.extractedSkills || []).find(
      s => s.name.toLowerCase() === reqSkill.name.toLowerCase()
    );
    const hasCert = candidateSkill?.certified || false;
    const resumeMentions = candidateSkill?.resumeMentions || (candidateSkill ? 1 : 0);

    let factor = 0.0;
    if (repoCount >= 2) {
      factor = 1.0;
      strengths.push({ skill: reqSkill.name, evidence: `${repoCount} GitHub repositories with verified code`, level: "Verified Strong Proof" });
    } else if (repoCount === 1) {
      factor = 0.85;
      strengths.push({ skill: reqSkill.name, evidence: "1 GitHub repository (Verified in Code)", level: "Verified Match" });
    } else if (hasCert) {
      factor = 0.55;
    } else if (resumeMentions > 0) {
      factor = 0.30;
      if (isRequired) {
        criticalGaps.push({ skill: reqSkill.name, reason: "Self-reported on resume with 0 verified code repositories", severity: "High" });
      }
    } else {
      factor = 0.0;
      if (isRequired) {
        criticalGaps.push({ skill: reqSkill.name, reason: "Not found in analyzed technical evidence", severity: "High" });
      }
    }

    earnedSkillScore += factor * skillWeight;
  });

  const skillAlignmentVal = totalSkillWeight > 0 
    ? Math.min(100, Math.round((earnedSkillScore / totalSkillWeight) * 100))
    : 0;

  // 2. GitHub Code Evidence (Weight: 25%)
  const repos = candidate.githubRepos || [];
  let codeEvidenceVal = 0;
  if (repos.length > 0) {
    const volumeScore = Math.min(40, repos.length * 10);
    const avgComplexity = repos.reduce((acc, r) => acc + (r.complexityScore || r.complexity_score || 5), 0) / repos.length;
    const complexityScore = Math.min(35, (avgComplexity / 10) * 35);
    const totalCommits = repos.reduce((acc, r) => acc + (r.commits || r.commits_last_year || 0), 0);
    const commitScore = Math.min(15, (totalCommits / 150) * 15);
    const docScore = 10; // baseline 800 words
    codeEvidenceVal = Math.min(100, Math.round(volumeScore + complexityScore + commitScore + docScore));
  }

  // 3. Production Hygiene (Weight: 20%)
  let productionHygieneVal = 0;
  if (repos.length > 0) {
    let testPoints = 0;
    repos.forEach(r => {
      const cov = r.testCoverage ?? r.test_coverage;
      const hasTests = r.hasTests || r.has_tests;
      if (cov !== null && cov !== undefined) {
        testPoints += cov > 0 ? (cov / 100) * 35 : 5;
      } else if (hasTests) {
        testPoints += 17.5; // baseline presence
      }
    });
    const testScore = testPoints / repos.length;
    const dockerScore = (repos.filter(r => r.hasDocker || r.has_docker).length / repos.length) * 35;
    const ciScore = (repos.filter(r => r.hasCI || r.has_ci).length / repos.length) * 30;
    productionHygieneVal = Math.min(100, Math.round(testScore + dockerScore + ciScore));
  }

  // 4. Architecture Depth (Weight: 20%)
  let architectureDepthVal = 0;
  if (repos.length > 0) {
    let score = 0;
    if (repos.some(r => r.hasModularStructure || r.has_modular_structure)) score += 25;
    if (repos.some(r => r.hasDbIntegration || r.has_db_integration)) score += 20;
    if (repos.some(r => r.hasEnvConfig || r.has_env_config)) score += 15;
    if (repos.some(r => r.hasDocker || r.has_docker)) score += 20;
    if (repos.some(r => (r.techStack || r.tech_stack || []).some(t => ['FastAPI', 'REST APIs', 'GraphQL', 'Django'].includes(t)))) score += 20;
    architectureDepthVal = Math.min(100, score);
  }

  // Composite Deterministic Score
  const baseScore = Math.min(100, Math.max(0, Math.round(
    skillAlignmentVal * 0.35 +
    codeEvidenceVal * 0.25 +
    productionHygieneVal * 0.20 +
    architectureDepthVal * 0.20
  )));

  // Milestone Progress Bonus for Simulator View
  let milestoneBonus = 0;
  if (completedMilestoneIds && completedMilestoneIds.length > 0) {
    completedMilestoneIds.forEach(id => {
      if (id === "build-fastapi-service") milestoneBonus += 14;
      if (id === "dockerize-repos") milestoneBonus += 8;
      if (id === "add-pytest-ci") milestoneBonus += 6;
      if (id === "aws-deploy-ecs") milestoneBonus += 5;
      if (id === "redis-caching-layer") milestoneBonus += 9;
      if (id === "k8s-helm-setup") milestoneBonus += 11;
      if (id === "dsa-graph-mastery") milestoneBonus += 4;
    });
  }

  // Rating Thresholds
  let rating = "Developing Readiness";
  let statusBadge = "Developing Readiness";
  if (baseScore >= 85) {
    rating = "High Readiness";
    statusBadge = "High Readiness";
  } else if (baseScore >= 70) {
    rating = "Moderate Readiness";
    statusBadge = "Moderate Readiness";
  } else if (baseScore >= 50) {
    rating = "Developing Readiness";
    statusBadge = "Developing Readiness";
  } else {
    rating = "Emerging Readiness";
    statusBadge = "Emerging Readiness";
  }

  return {
    overallScore: baseScore,
    baseScore: baseScore,
    milestoneBonus: milestoneBonus,
    pillars: {
      skillAlignment: skillAlignmentVal,
      codeEvidence: codeEvidenceVal,
      productionHygiene: productionHygieneVal,
      architectureDepth: architectureDepthVal
    },
    rating,
    statusBadge,
    strengths,
    criticalGaps
  };
}
